plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts

```diff
@@ -0,0 +1,249 @@
+ * limitations under the License.
+ */
+
+import { Entity, stringifyEntityRef } from '@backstage/catalog-model';
+import {
+  CatalogProcessor,
+  CatalogProcessorCache,
+} from '@backstage/plugin-catalog-node';
+import { DiscoveryService } from '@backstage/backend-plugin-api';
+import { Languages } from '@backstage/plugin-linguist-common';
+import fetch from 'node-fetch';
+import { Logger } from 'winston';
+import { HumanDuration, durationToMilliseconds } from '@backstage/types';
+
+/**
+ * A function which takes a language string returned by linguist and
+ * cleans it up prior to exporting to entity tags. If the language is emitted as
+ * the "undefined" type, then it will not be added as a tag
+ * @public
+ */
+export type TagTransformer = (tag: string) => string | undefined;
+/**
+ * A function which given an entity, determines if it should be processed for linguist tags.
+ * @public
+ */
+export type ShouldProcessEntity = (entity: Entity) => boolean;
+
+interface CachedData {
+  [key: string]: number | string[];
+  languages: string[];
+  cachedTime: number;
+}
+
+/**
+ * The constructor arguments for building the  LinguistTagsProcessor
+ * @public
+ */
+export interface LinguistTagsProcessorArgs {
+  logger: Logger;
+  discovery: DiscoveryService;
+  /**
+   * Optional transformer function which can convert a language tag prior to being
+   * attached to an entity. If the tag transformer emits an "undefined" type, then it
+   * will not be added as a tag to the entity.
+   */
+  tagTransformer?: TagTransformer;
+  /**
+   * A function which determines which entities should be processed by the LinguistTagProcessor.
+   *
+   * The default is to process all entities of kind=Component
+   */
+  shouldProcessEntity?: ShouldProcessEntity;
+  /**
+   * Determines how long to cache language breakdowns for entities in the processor. Considering
+   * how often this processor runs, caching can help move some read traffic off of the linguist DB.
+   *
+   * If this caching is using up too much memory, you can disable it by setting cacheTTL to 0.
+   */
+  cacheTTL?: HumanDuration;
+  /**
+   * How many bytes must exist of a language in a repo before we consider it for adding a tag to
+   * the entity. This can be used if some repos have short utility scripts that may not be the primary
+   * language for the repo.
+   */
+  bytesThreshold?: number;
+}
+
+/**
+ * These are customized transformers for certain languages where
+ * regex is not good enough to clean them up.
+ */
+const CUSTOM_LANG_MAP: Record<string, string> = {
+  'asp.net': 'asp-dot-net',
+};
+
+/**
+ * Converts language tags from linguist to something acceptable by
+ * the tag requirements for backstage
+ *
+ * @param tag - A language tag from linguist
+ * @returns Cleaned up language tag
+ * @public
+ */
+export const defaultTagTransformer: TagTransformer = (tag: string) => {
+  const cleaned = tag.replace(/\s+/g, '-').toLowerCase();
+  return CUSTOM_LANG_MAP[cleaned] || cleaned;
```

Comment:
I intentionally chose to use `||` operator here. I don't see a point in building a custom map to any falsey value, that would be an error in my mind.

As for checking the value in the map before doing the transforms, I don't have a strong opinion one way or another. As it is now, the keys in the `CUSTOM_LANG_MAP` are already transformed, which is why the mapping is `asp.net => asp-dot-net`. So in this case, it is correct to pass in `CUSTOM_LANG_MAP[cleaned]`.

I do wonder if there should just be a global replace of `.` to `-dot-`. In your example of replacers above, they just get stripped out entirely. Is that better?

Reply:
sure, makes sense to keep `||` then. Just wanted to confirm.

Re `CUSTOM_LANG_MAP[cleaned]`, you are right, your code worked. I just misread/thought you would replace more already.

However, I would still recommend sticking to the original value instead of the already mutated one.
You can use the actual language names from linguist as keys then. To me that seems far more easier to maintain and understand.

In this case, `ASP.NET` instead of `asp.net`.

Also, the static mapping makes the standard transformation/cleaning unnecessary. And we can even skip the execution of the regex replace operation(s).

As there is just one exception at the moment, the difference will be subtle, though.

Reply:
> In your example of replacers above, they just get stripped out entirely. Is that better?

As long as there is no other mapping, yes, they would get removed. All invalid chars will be removed/replaced. They are not allowed characters. If special handling for certain chars is desired, it could be added on top.


> I do wonder if there should just be a global replace of . to -dot-. 

So far, there are the following ones:
```ts
['ASN.1', 'ASP.NET', 'Gemfile.lock', 'PEG.js', 'Visual Basic .NET', 'Visual Basic 6.0', 'robots.txt']
```

if we apply this as general rule, we end up with
```ts
['asn-dot-1', 'asp-dot-net', 'gemfile-dot-lock', 'peg-dot-js', 'visual-basic-dot-net', 'visual-Basic-6-dot-0', 'robots-dot-txt']
```
vs
```ts
['asn-1', 'asp-net', 'gemfile-lock', 'peg-js', 'visual-basic-net', 'visual-basic-6-0', 'robots-txt']
```

Personally, I think it only makes sense for `ASP.NET` and probably also `Visual Basic .NET`, at which you speak `.` as `dot` as well.

Basically, `.NET` as `-dot-net` could make sense.

Reply:
I took your suggestions and am only replacing a custom rule for `.net` specifically. I think that makes sense.

I also took a slightly different approach with regex which I think is a bit cleaner / easier to read / and handles cases of double slashes.

Also added tests across the full linguist data-set, including a jest snapshot which can now be easily reviewed when making transformation changes / linguist adding new supported languages.